### PR TITLE
Create UI List genres  and handle click in item and button more

### PR DESCRIPTION
**Pull request by NguyenDinhTrong99**
## [Pull request title]
 - Create UI item
 - Create adapter
 - Receive data from presenter and set to adapter
 - create interface handle click 'more'
 - handle click icon more in home screen
### [Redmine link]
https://edu-redmine.sun-asterisk.vn/issues/17285
#### Screenshots:
Screenshot_1568895798

### CODE REVIEW

File: app/src/main/java/com/trongdeptrai/soundcloud/screen/home/adapter/GenresAdapter.java [hunk 1]
```diff
@@ -99,18 +114,44 @@ void bindView() {
...
         }
-    }
 
-    public void updateData(List<Genre> data) {
-        mGenres.clear();
-        mGenres.addAll(data);
-        notifyItemInserted(mGenres.size());
+        @Override
+        public void onClick(View view) {
+            mOnClickItemMoreListener.onClickMore(mGenres.get(getAdapterPosition()));
```
Comment (thaohq-0814) [inline app/src/main/java/com/trongdeptrai/soundcloud/screen/home/adapter/GenresAdapter.java]: check null

File: app/src/main/java/com/trongdeptrai/soundcloud/screen/home/adapter/TrackAdapter.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+public class TrackAdapter extends RecyclerView.Adapter<TrackAdapter.ViewHolder> {
+    private List<Track> mTracks;
+    private OnItemRecyclerViewClickListener<List<Track>, Track>
+            mTrackOnItemRecyclerViewClickListener;
+
+    TrackAdapter(List<Track> tracks) {
+        mTracks = tracks;
+    }
+
+    void setTrackOnItemRecyclerViewClickListener(
```
Comment (thaohq-0814) [inline app/src/main/java/com/trongdeptrai/soundcloud/screen/home/adapter/TrackAdapter.java]:
```suggestion
    void setItemClickListener(
```

File: app/src/main/java/com/trongdeptrai/soundcloud/screen/home/adapter/TrackAdapter.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import com.trongdeptrai.soundcloud.utils.OnItemRecyclerViewClickListener;
+import java.util.List;
+
+import static com.trongdeptrai.soundcloud.utils.Common.getBigImageUrl;
+import static com.trongdeptrai.soundcloud.utils.Common.getImageUrl;
+
+public class TrackAdapter extends RecyclerView.Adapter<TrackAdapter.ViewHolder> {
+    private List<Track> mTracks;
+    private OnItemRecyclerViewClickListener<List<Track>, Track>
+            mTrackOnItemRecyclerViewClickListener;
```
Comment (thaohq-0814) [inline app/src/main/java/com/trongdeptrai/soundcloud/screen/home/adapter/TrackAdapter.java]:
```suggestion
            mItemClickListener;
```

File: app/src/main/java/com/trongdeptrai/soundcloud/screen/home/adapter/TrackAdapter.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @Override
+    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
+        holder.bindView();
+    }
+
+    @Override
+    public int getItemCount() {
+        return mTracks.size();
```
Comment (thaohq-0814) [inline app/src/main/java/com/trongdeptrai/soundcloud/screen/home/adapter/TrackAdapter.java]:
```suggestion
        return mTracks  != null ? mTrack.size() : 0;
```

File: app/src/main/java/com/trongdeptrai/soundcloud/screen/home/adapter/GenresAdapter.java [hunk 5]
```diff
@@ -146,11 +146,12 @@ void bindView() {
...
             TrackAdapter adapter = new TrackAdapter(mGenres.get(getAdapterPosition()).getTracks());
             mRecyclerViewType.setAdapter(adapter);
-            adapter.setTrackOnItemRecyclerViewClickListener(mOnItemRecyclerViewClickListener);
+            adapter.setItemClickListener(mOnItemRecyclerViewClickListener);
         }
 
         @Override
         public void onClick(View view) {
+            if(mOnClickItemMoreListener != null)
             mOnClickItemMoreListener.onClickMore(mGenres.get(getAdapterPosition()));
```
Comment (thaohq-0814) [inline app/src/main/java/com/trongdeptrai/soundcloud/screen/home/adapter/GenresAdapter.java]: format code

**Review by thaohq-0814 — APPROVED** [review]

File: app/src/main/java/com/trongdeptrai/soundcloud/screen/home/HomeFragment.java [hunk 6]
```diff
@@ -21,8 +21,9 @@
 import java.util.ArrayList;
 import java.util.List;
 
-public class HomeFragment extends BaseFragment implements HomeContract.View,
-        OnItemRecyclerViewClickListener<List<Track>, Track> {
+public class HomeFragment extends BaseFragment
+        implements HomeContract.View, OnItemRecyclerViewClickListener<List<Track>, Track>,
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/trongdeptrai/soundcloud/screen/home/HomeFragment.java]: tại sao onclick lại truyền 1 list và 1 object vậy e?

File: app/src/main/java/com/trongdeptrai/soundcloud/screen/home/adapter/TrackAdapter.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            if (track.getArtworkUrl() != null) {
+                getImageUrl(getBigImageUrl(track.getArtworkUrl()), this);
+            } else {
+                onGetImageFailed();
+            }
+        }
+
+        @Override
+        public void onClick(View view) {
+            mListener.onItemClickListener(mTracks, mTracks.get(getAdapterPosition()));
```
Comment (tuanlvt-1117) [inline app/src/main/java/com/trongdeptrai/soundcloud/screen/home/adapter/TrackAdapter.java]:
```suggestion
            mListener.onItemClickListener(mTracks, getAdapterPosition());
```

**Review by tuanlvt-1117 — APPROVED** [review]

**Pull request merged**